src/Platform.Handlers/src/Xamarin.Platform.Handlers/Hosting/AppBuilder.cs

```diff
@@ -0,0 +1,271 @@
+			ServiceCollection servicesCollection = new ServiceCollection();
+
+			var host = Build();
+
+			if (_handlersServiceProvider != null)
+				servicesCollection.AddSingleton<IHandlerServiceProvider>(_handlersServiceProvider);
+
+			var app = (TApplication)Activator.CreateInstance(typeof(TApplication));
+
+			if (_hostBuilderContext != null)
+				AppLoader.ConfigureAppServices<TApplication>(_hostBuilderContext, servicesCollection, app);
+
+			var services = servicesCollection.BuildServiceProvider();
+
+			(app as App)?.SetServiceProvider(services);
+
+			return (host, app);
+		}
+
+		public IHost Build()
+		{
+			if (_hostBuilt)
+			{
+				throw new InvalidOperationException("Build can only be called once.");
+			}
+			_hostBuilt = true;
+
+			// the order is important here
+			BuildHostConfiguration();
+			CreateHostingEnvironment();
+			CreateHostBuilderContext();
+			BuildAppConfiguration();
+			CreateHandlersServiceProvider();
+			CreateServiceProvider();
+			if (_appServices == null)
+			{
+				throw new InvalidOperationException($"The (_appServices) cannot be null");
+			}
+			return _appServices.GetRequiredService<IHost>();
+		}
+
+		public IHostBuilder ConfigureAppConfiguration(Action<HostBuilderContext, IConfigurationBuilder> configureDelegate)
+		{
+			_configureAppConfigActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+			return this;
+		}
+
+		public IHostBuilder ConfigureContainer<TContainerBuilder>(Action<HostBuilderContext, TContainerBuilder> configureDelegate)
+		{
+			_configureContainerActions.Add(new ConfigureContainerAdapter<TContainerBuilder>(configureDelegate
+			 ?? throw new ArgumentNullException(nameof(configureDelegate))));
+			return this;
+		}
+
+		public IHostBuilder ConfigureHostConfiguration(Action<IConfigurationBuilder> configureDelegate)
+		{
+			_configureHostConfigActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+			return this;
+		}
+
+		public IHostBuilder ConfigureServices(Action<HostBuilderContext, IServiceCollection> configureDelegate)
+		{
+			_configureServicesActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+			return this;
+		}
+
+		public IHostBuilder ConfigureHandlers(Action<HostBuilderContext, IHandlerServiceCollection> configureDelegate)
+		{
+			_configureHandlersActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+			return this;
+		}
+
+#pragma warning disable CS8714 // The type cannot be used as type parameter in the generic type or method. Nullability of type argument doesn't match 'notnull' constraint.
+		public IHostBuilder UseServiceProviderFactory<TContainerBuilder>(IServiceProviderFactory<TContainerBuilder> factory)
+		{
+			if (factory != null)
+				_serviceProviderFactory = new ServiceFactoryAdapter<TContainerBuilder>(factory ?? throw new ArgumentNullException(nameof(factory)));
```

Comment:
Why the double check for `factory` being null? The line above checks that it is not null, so the `?? throw new ArgumentNullException` is dead code. Same applies in the method below.